src/lib.rs

```diff
@@ -1,8 +1,526 @@
+                    let json = Derive::all("JsonSchema");
+
+                    if !self.derive_traits.contains(&json) {
+                        self.derive_traits.push(json)
+                    }
+                }
+                _ => {}
+            }
+        }
+
+        /// Add the supplied [`Derive`] directive to the list of traits to be derived on
+        /// generated types
+        pub fn derive(&mut self, value: Derive) {
+            self.derive_traits.push(value);
+        }
+
+        /// Attempt to parse the supplied value as a [`Derive`] directive and add it to
+        /// the list of traits to be derived on generated types
+        pub fn try_derive(&mut self, value: impl AsRef<str>) {
+            let derive = value.as_ref().parse::<Derive>().expect("unparsable value");
+            self.derive_traits.push(derive);
+        }
+
+        /// Add all the supplied [`Derive`] directives to the list of traits to be derived
+        /// on generated types
+        pub fn derive_all(&mut self, values: impl IntoIterator<Item = Derive>) {
+            self.derive_traits.extend(values);
+        }
+
+        /// Attempt to parse each of the supplied values as a [`Derive`] directive and add
+        /// them to the list of traits to be derived on generated types
+        pub fn try_derive_all(&mut self, values: impl IntoIterator<Item = impl AsRef<str>>) {
+            let values = values.into_iter().map(|value| {
+                value.as_ref().parse::<Derive>().expect("unparsable value")
+            });
+
+            self.derive_traits.extend(values);
+        }
+    )
+)]
+pub struct KopiumTypeGenerator {
+    /// Use this CRD version if multiple versions are present
+    api_version: Option<String>,
+
+    /// Do not emit prelude(s)
+    hide_prelude: bool,
+
+    /// Do not derive CustomResource nor set kube-derive attributes
+    ///
+    /// If this is set, it makes any kube-derive specific options such as `--schema` unnecessary
+    hide_kube: bool,
+
+    /// Emit doc comments from descriptions
+    #[builder(via_mutators)]
+    emit_docs: bool,
+
+    /// Emit builder derives via the typed_builder crate
+    builders: bool,
+
+    /// Schema mode to use for kube-derive
+    ///
+    /// The default is --schema=disabled and will compile without a schema,
+    /// but the resulting crd cannot be applied into a cluster.
+    ///
+    /// --schema=manual requires the user to `impl JsonSchema for MyCrdSpec` elsewhere for the code to compile.
+    /// Once this is done, the crd via `CustomResourceExt::crd()` can be applied into Kubernetes directly.
+    ///
+    /// --schema=derived implies `--derive JsonSchema`. The resulting schema will compile without external user action.
+    /// The crd via `CustomResourceExt::crd()` can be applied into Kubernetes directly.
+    #[builder(
+        default_code = r#"String::from("disabled")"#,
+        via_mutators(init = String::from("disabled")),
+    )]
+    schema_mode: String,
```

Comment:
is the builder code like this flexible? the suggestion i made may be not super compatible with an inflexible builder setup?

Reply:
In my fiddling so far, it has been, yeah. It's really only there to try and maintain parity with the CLI behavior though, which may or may not be sensible?

Reply:
i do have a stronger sense of guarantees for library behavior baked into my head personally, because you can do a lot more of the common rust mantra (make invalid states unrepresentable) with the code as opposed to various cli flags that ultimately takes strings.

we don't have to go too far with it, in an initial pr like this, but ideally we can make the library interface follow standard library conventions a little more. as in, i don't think we need to maintain parity with the ad-hoc way i set up most of the kopium cli flags.

Reply:
Well, when you put it like that, I think it makes a lot of sense. Seems like we could go the other way with it actually, and start (in a follow-up) by bringing the library interface up to snuff with standard rust library conventions and then bringing the CLI into line so to speak.

(Side note, please take *all* of ^ these type comments from me as a tacit offer to do the work in question, as opposed to "wouldn't it be great if someone..." 😅)